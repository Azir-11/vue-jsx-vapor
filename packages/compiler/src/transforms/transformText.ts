import {
  type BlockIRNode,
  DynamicFlag,
  IRNodeTypes,
  type RootNode,
} from '../ir'
import {
  getLiteralExpressionValue,
  isEmptyText,
  isJSXComponent,
  resolveExpression,
  resolveJSXText,
} from '../utils'
import { processConditionalExpression, processLogicalExpression } from './vIf'
import type { NodeTransform, TransformContext } from '../transform'
import type {
  JSXElement,
  JSXExpressionContainer,
  JSXText,
  Node,
} from '@babel/types'

type TextLike = JSXText | JSXExpressionContainer
const seen = new WeakMap<
  TransformContext<RootNode>,
  WeakSet<TextLike | BlockIRNode['node'] | RootNode>
>()

export const transformText: NodeTransform = (node, context) => {
  if (!seen.has(context.root)) seen.set(context.root, new WeakSet())
  if (seen.get(context.root)!.has(node)) {
    context.dynamic.flags |= DynamicFlag.NON_TEMPLATE
    return
  }

  if (
    node.type === 'JSXElement' &&
    !(isJSXComponent(node) as boolean) &&
    isAllTextLike(node.children)
  ) {
    processTextLikeContainer(
      node.children,
      context as TransformContext<JSXElement>,
    )
  } else if (node.type === 'JSXExpressionContainer') {
    if (node.expression.type === 'ConditionalExpression') {
      return processConditionalExpression(node.expression, context)
    } else if (node.expression.type === 'LogicalExpression') {
      return processLogicalExpression(node.expression, context)
    } else {
      processTextLike(context as TransformContext<JSXExpressionContainer>)
    }
  } else if (node.type === 'JSXText') {
    const value = resolveJSXText(node)
    if (value) {
      context.template += value
    } else {
      context.dynamic.flags |= DynamicFlag.NON_TEMPLATE
    }
  }
}

function processTextLike(context: TransformContext<JSXExpressionContainer>) {
  const nexts = context.parent!.node.children?.slice(context.index)
  const idx = nexts.findIndex((n) => !isTextLike(n))
  const nodes = (idx > -1 ? nexts.slice(0, idx) : nexts) as Array<TextLike>

  const values = createTextLikeExpressions(nodes, context)
  if (!values.length) return

  context.dynamic.flags |= DynamicFlag.INSERT | DynamicFlag.NON_TEMPLATE
  context.registerOperation({
    type: IRNodeTypes.CREATE_TEXT_NODE,
    id: context.reference(),
    values,
    effect: false,
  })
}

function processTextLikeContainer(
  children: TextLike[],
  context: TransformContext<JSXElement>,
) {
  const values = createTextLikeExpressions(children, context)
  if (!values.length) return

  const literals = values.map(getLiteralExpressionValue)
  if (literals.every((l) => l != null)) {
    context.childrenTemplate = literals.map((l) => String(l))
  } else {
    context.registerOperation({
      type: IRNodeTypes.SET_TEXT,
      element: context.reference(),
      values,
    })
  }
}

function createTextLikeExpressions(
  nodes: TextLike[],
  context: TransformContext,
) {
  const values = []
  for (const node of nodes) {
    if (isEmptyText(node)) continue
    seen.get(context.root)!.add(node)
    values.push(resolveExpression(node, context, true))
  }
  return values
}

function isAllTextLike(children: Node[]): children is TextLike[] {
  return (
    !!children.length &&
    children.every(isTextLike) &&
    // at least one an interpolation
    children.some((n) => n.type === 'JSXExpressionContainer')
  )
}

function isTextLike(node: Node): node is TextLike {
  return (
    (node.type === 'JSXExpressionContainer' &&
      !(
        node.expression.type === 'ConditionalExpression' ||
        node.expression.type === 'LogicalExpression'
      )) ||
    node.type === 'JSXText'
  )
}
