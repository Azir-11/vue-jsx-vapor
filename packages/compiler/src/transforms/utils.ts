import {
  DynamicFlag,
  type IRDynamicInfo,
  type IRNodeTypes,
} from '@vue/compiler-vapor'
import { createSimpleExpression } from '@vue/compiler-dom'
import {
  type Expression,
  type JSXElement,
  type JSXFragment,
  arrowFunctionExpression,
  callExpression,
  jsxClosingFragment,
  jsxExpressionContainer,
  jsxFragment,
  jsxOpeningFragment,
  parenthesizedExpression,
} from '@babel/types'
import { isJSXElement } from '../utils'
import type { TransformContext } from '../transform'
import type { BlockIRNode } from '../ir/index'

export function newDynamic(): IRDynamicInfo {
  return {
    flags: DynamicFlag.REFERENCED,
    children: [],
  }
}

export function newBlock(node: BlockIRNode['node']): BlockIRNode {
  return {
    type: 1 satisfies IRNodeTypes.BLOCK,
    node,
    dynamic: newDynamic(),
    effect: [],
    operation: [],
    returns: [],
  }
}

export function createBranch(
  node: Parameters<typeof wrapFragment>[0],
  context: TransformContext,
  isVFor?: boolean,
): [BlockIRNode, () => void] {
  context.node = node = wrapFragment(node, isVFor)

  const branch: BlockIRNode = newBlock(node)
  const exitBlock = context.enterBlock(branch, isVFor)
  context.reference()
  return [branch, exitBlock]
}

export function wrapFragment(
  node: JSXElement | JSXFragment | Expression,
  isVFor?: boolean,
): JSXFragment {
  if (node.type === 'JSXFragment') {
    return node
  }

  if (
    isVFor &&
    (node.type === 'ArrowFunctionExpression' ||
      node.type === 'FunctionExpression')
  ) {
    if (isJSXElement(node.body)) {
      node = node.body
    } else if (
      node.body.type === 'BlockStatement' &&
      node.body.body[0].type === 'ReturnStatement' &&
      node.body.body[0].argument
    ) {
      node = node.body.body[0].argument
    } else {
      node = {
        ...callExpression(
          parenthesizedExpression(arrowFunctionExpression([], node.body)),
          [],
        ),
        start: node.body.start,
        end: node.body.end,
      }
    }
  }

  return jsxFragment(jsxOpeningFragment(), jsxClosingFragment(), [
    node.type === 'JSXElement' ? node : jsxExpressionContainer(node),
  ])
}

export const EMPTY_EXPRESSION = createSimpleExpression('', true)
