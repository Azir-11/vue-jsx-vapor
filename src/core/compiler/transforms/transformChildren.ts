import { DynamicFlag, type IRDynamicInfo, IRNodeTypes } from '../ir/index'
import {
  type NodeTransform,
  type TransformContext,
  transformNode,
} from '../transform'
import type { Node } from '@babel/types'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT || node.type === 'JSXFragment'

  if (!isFragment && node.type !== 'JSXElement') return

  Array.from(node.children).forEach((child, index) => {
    if (child.type === 'JSXText' && !child.value.trim()) {
      child.value = ' '
      if (!index) {
        node.children.splice(0, 1)
      } else if (index === node.children.length) {
        node.children.splice(-1, 1)
      }
    }
  })

  for (const [i, child] of node.children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<Node>)
  }
}

function processDynamicChildren(context: TransformContext<Node>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`

          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())

          context.registerOperation({
            type: IRNodeTypes.INSERT_NODE,
            elements: prevDynamics.map((child) => child.id!),
            parent: context.reference(),
            anchor,
          })
        } else {
          context.registerOperation({
            type: IRNodeTypes.PREPEND_NODE,
            elements: prevDynamics.map((child) => child.id!),
            parent: context.reference(),
          })
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length) {
    context.registerOperation({
      type: IRNodeTypes.INSERT_NODE,
      elements: prevDynamics.map((child) => child.id!),
      parent: context.reference(),
    })
  }
}
