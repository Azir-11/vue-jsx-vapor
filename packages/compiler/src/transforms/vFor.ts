import {
  ErrorCodes,
  type SimpleExpressionNode,
  createCompilerError,
  forAliasRE,
  isConstantNode,
} from '@vue/compiler-dom'
import { parseExpression } from '@babel/parser'
import { DynamicFlag, IRNodeTypes } from '../ir'
import {
  findProp,
  getText,
  isJSXComponent,
  propToExpression,
  resolveExpression,
  resolveLocation,
  resolveSimpleExpression,
} from '../utils'
import {
  type NodeTransform,
  type TransformContext,
  createStructuralDirectiveTransform,
} from '../transform'
import { createBranch } from './utils'
import type { JSXAttribute, JSXElement, Node } from '@babel/types'

export const transformVFor: NodeTransform = createStructuralDirectiveTransform(
  'for',
  processFor,
)

export function processFor(
  node: JSXElement,
  dir: JSXAttribute,
  context: TransformContext,
) {
  if (!dir.value) {
    context.options.onError(
      createCompilerError(
        ErrorCodes.X_V_FOR_NO_EXPRESSION,
        resolveLocation(dir.loc, context),
      ),
    )
    return
  }
  if (!forAliasRE) {
    context.options.onError(
      createCompilerError(
        ErrorCodes.X_V_FOR_MALFORMED_EXPRESSION,
        resolveLocation(dir.loc, context),
      ),
    )
    return
  }

  let value: SimpleExpressionNode | undefined,
    index: SimpleExpressionNode | undefined,
    key: SimpleExpressionNode | undefined,
    source: SimpleExpressionNode
  if (
    dir.value.type === 'JSXExpressionContainer' &&
    dir.value.expression.type === 'BinaryExpression'
  ) {
    if (dir.value.expression.left.type === 'SequenceExpression') {
      const expressions = dir.value.expression.left.expressions
      value = expressions[0] && resolveValueExpression(expressions[0], context)
      key = expressions[1] && resolveExpression(expressions[1], context)
      index = expressions[2] && resolveExpression(expressions[2], context)
    } else {
      value = resolveValueExpression(dir.value.expression.left, context)
    }
    source = resolveExpression(dir.value.expression.right, context)
  }

  const keyProp = findProp(node, 'key')
  const keyProperty = keyProp && propToExpression(keyProp, context)
  const isComponent = isJSXComponent(node)
  const id = context.reference()
  context.dynamic.flags |= DynamicFlag.NON_TEMPLATE | DynamicFlag.INSERT
  const [render, exitBlock] = createBranch(node, context, true)
  return (): void => {
    exitBlock()

    const { parent } = context

    // if v-for is the only child of a parent element, it can go the fast path
    // when the entire list is emptied
    const isOnlyChild =
      parent &&
      parent.block.node !== parent.node &&
      parent.node.children.length === 1

    context.registerOperation({
      type: IRNodeTypes.FOR,
      id,
      source,
      value,
      key,
      index,
      keyProp: keyProperty,
      render,
      once: context.inVOnce || !!(source.ast && isConstantNode(source.ast, {})),
      component: isComponent,
      onlyChild: !!isOnlyChild,
    })
  }
}

function resolveValueExpression(node: Node, context: TransformContext) {
  const text = getText(node, context)
  return node.type === 'Identifier'
    ? resolveSimpleExpression(text, false, node.loc)
    : resolveSimpleExpression(
        text,
        false,
        node.loc,
        parseExpression(`(${text})=>{}`, {
          plugins: ['typescript'],
        }),
      )
}
