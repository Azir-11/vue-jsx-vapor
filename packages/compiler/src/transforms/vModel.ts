import { transformVModel as _transformVModel } from '@vue/compiler-vapor'
import { resolveDirectiveNode, resolveNode } from '../utils'
import type { DirectiveTransform } from '../transform'

export const transformVModel: DirectiveTransform = (dir, node, context) => {
  return _transformVModel(
    resolveDirectiveNode(dir, context),
    resolveNode(node, context),
    context as any,
  )
}
