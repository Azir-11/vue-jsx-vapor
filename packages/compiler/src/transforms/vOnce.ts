import { NodeTypes, findDir } from '@vue/compiler-dom'
import { findProp } from '../utils'
import type { NodeTransform } from '../transform'

export const transformVOnce: NodeTransform = (node, context) => {
  if (
    // !context.inSSR &&
    node.type === 'JSXElement' &&
    findProp(node, 'v-once')
  ) {
    context.inVOnce = true
  }
}
