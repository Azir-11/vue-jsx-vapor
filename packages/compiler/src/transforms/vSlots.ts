import { createCompilerError, ErrorCodes } from '@vue/compiler-dom'
import { IRSlotType } from '../ir'
import { isJSXComponent, resolveExpression, resolveLocation } from '../utils'
import type { NodeTransform } from '../transform'
import type { JSXAttribute } from '@babel/types'

export const transformVSlots: NodeTransform = (node, context) => {
  if (node.type !== 'JSXElement' || !isJSXComponent(node)) return

  const {
    openingElement: { attributes },
    children,
  } = node

  const vSlotsIndex = attributes.findIndex(
    (attr) =>
      attr.type === 'JSXAttribute' && attr.name.name.toString() === 'v-slots',
  )
  const vSlotsDir = attributes[vSlotsIndex] as JSXAttribute
  if (vSlotsDir && vSlotsDir.value?.type === 'JSXExpressionContainer') {
    attributes.splice(vSlotsIndex, 1)
    context.slots = [
      {
        slotType: IRSlotType.EXPRESSION,
        slots: resolveExpression(vSlotsDir.value.expression, context),
      },
    ]

    if (children.length) {
      context.options.onError(
        createCompilerError(
          ErrorCodes.X_V_SLOT_MIXED_SLOT_USAGE,
          resolveLocation(children[0].loc, context),
        ),
      )
    }
  }
}
