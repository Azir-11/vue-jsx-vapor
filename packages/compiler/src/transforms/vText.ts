import { transformVText as _transformVText } from '@vue/compiler-vapor'
import { resolveDirectiveNode, resolveNode } from '../utils'
import type { DirectiveTransform } from '../transform'

export const transformVText: DirectiveTransform = (dir, node, context) => {
  return _transformVText(
    resolveDirectiveNode(dir, context),
    resolveNode(node, context),
    context as any,
  )
}
