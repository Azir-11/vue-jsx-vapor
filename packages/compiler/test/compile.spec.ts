import { describe, expect, test } from 'vitest'
import {
  transformChildren,
  transformElement,
  transformTemplateRef,
  transformText,
  transformVBind,
  transformVHtml,
  transformVModel,
  transformVOn,
  transformVShow,
  transformVSlot,
} from '../src'
import { makeCompile } from './transforms/_utils'

const compile = makeCompile({
  nodeTransforms: [
    transformTemplateRef,
    transformText,
    transformElement,
    transformVSlot,
    transformChildren,
  ],
  directiveTransforms: {
    bind: transformVBind,
    on: transformVOn,
    model: transformVModel,
    show: transformVShow,
    html: transformVHtml,
  },
})

describe('compile', () => {
  test('static template', () => {
    const { code } = compile(
      `<div>
        <div>hello</div>
        <input />
        <span />
      </div>`,
    )
    expect(code).toMatchInlineSnapshot(`
      "
        const n0 = t0()
        return n0
      "
    `)
  })

  test('dynamic root', () => {
    const { code } = compile(`<>{ 1 }{ 2 }</>`)
    expect(code).toMatchInlineSnapshot(`
      "
        const n0 = _createTextNode(1, 2)
        return n0
      "
    `)
  })

  test('dynamic root', () => {
    const { code } = compile(`<div>{a +b +       c }</div>`)
    expect(code).toMatchInlineSnapshot(`
      "
        const n0 = t0()
        _setText(n0, () => (a +b +       c))
        return n0
      "
    `)
  })

  describe('expression parsing', () => {
    test('interpolation', () => {
      const { code } = compile(`<>{ a + b }</>`, {
        inline: true,
      })
      expect(code).toMatchInlineSnapshot(`
        "
          const n0 = _createTextNode(() => (a + b))
          return n0
        "
      `)
      expect(code).contains('a + b')
    })
  })
})
