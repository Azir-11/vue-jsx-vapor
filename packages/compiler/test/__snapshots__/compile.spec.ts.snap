// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compile > setInsertionState > next, child and nthChild should be above the setInsertionState 1`] = `
"
  const n3 = t0()
  const n2 = _next(_child(n3))
  const p0 = _nthChild(n3, 4)
  const n1 = _child(p0)
  _setInsertionState(n3, n2)
  const n0 = _createComponent(Comp)
  _renderEffect(() => _setProp(n1, "disabled", foo))
  return n3
"
`;
